### PR TITLE
Add LFRic small matrix vector multiplication benchmark

### PR DESCRIPTION
**Pull request by sergisiso**
Ported from the benchmark developed for the IPCC project and updated README and LFRic infrastructure file when possible to the latest versions in https://github.com/christophermaynard/LFRic-microbenchmark/common

### CODE REVIEW

**Comment by sergisiso** [comment]
@arporter @rupertford @christophermaynard In this PR I introduced the LFRic mv benchmark created in the IPCC project as a new PSycloneBench benchmark. It is (mostly) as is, I just updated LFRic infrastructure files for the new ones using the modern license and updated of READMEs and Makefiles to reflect the new structure.

The large number of modified files in the PR is due to the renaming of a base directory.

**Comment by sergisiso** [comment]
I converted this to a draft PR to avoid being merged until I make sure git reflects the license change are taken from new external files, and not re-licensed them myself.

**Comment by sergisiso** [comment]
I have now modified the dino lib 'after' copying in the BSD-style licensed dino_mod.F90 and make the attributions clear in the [readme](https://github.com/stfc/PSycloneBench/tree/initial_matvec/benchmarks/smallmatvec/manual_versions/openmp).

I mark this PR ready to Review as i think now this can be safely merged.

File: benchmarks/smallmatvec/manual_versions/openmp/README.md [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+It was initially created by Chris Maynard using the PSyKE and Dino tools to
+extract the needed infrastructure and data from LFRic. Then extended by
+Sergi Siso to make it configurable to multiple OpenMP implementations and
+benchmark the KNL architecture for the IPCC at Hartree Centre project.
+
+Some files from the LFRic infrastructure are needed to run the benchmark,
+these are `argument_mod.F90`, `constants_mod.F90`, `dino_mod.F90` and
+`kernel_mod.F90`. These files have been copied from
+https://github.com/christophermaynard/LFRic-microbenchmarks/common
+and are distributed with the attached licence file LICENNCE.MetOffice.
```
Comment (arporter) [inline benchmarks/smallmatvec/manual_versions/openmp/README.md:32]: s/LICENNCE/LICENCE/

File: benchmarks/smallmatvec/manual_versions/openmp/README.md [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+scalability analysis made in the IPCC project.
+
+# Limitations
+
+Currently compiler and compiler flags are hardcoded into the Makefile, the
+PSycloneBench compiler environment variables are ignored.
+
+# Attributions 
+
+It was initially created by Chris Maynard using the PSyKE and Dino tools to
```
Comment (arporter) [inline benchmarks/smallmatvec/manual_versions/openmp/README.md:23]: Possibly give affiliations e.g. "Chris Maynard (UK Met Office)"

File: benchmarks/smallmatvec/manual_versions/openmp/innerloop.sh [hunk 3]
```diff
@@ -0,0 +1,16 @@
+#!/usr/bin/env bash
+
+output=innerloop.txt
```
Comment (arporter) [inline benchmarks/smallmatvec/manual_versions/openmp/innerloop.sh:3]: Please add a comment line explaining what this file does.

File: benchmarks/smallmatvec/manual_versions/openmp/kdriver.F90 [hunk 4]
```diff
@@ -0,0 +1,669 @@
...
+  write(*,*) "Colors=", ncolour, " -> ", ncp_colour
+  write(*,*) "Layers=", nlayers
+  write(*,*) "Horitzontal Cells=", ncell_3d/nlayers, ncell
+  write(*,*) "3D Cells=", ncell_3d
+  write(*,*) "ndf1=", ndf_any_space_1_theta_adv_term, ", undf1=", undf_any_space_1_theta_adv_term
+  write(*,*) "ndf2=", ndf_any_space_2_x, ", undf2=", undf_any_space_2_x
+
+
+  if (dotoutput.eq."write")then
+      open(1234, file="2dgrid.dot", action="write")
```
Comment (arporter) [inline benchmarks/smallmatvec/manual_versions/openmp/kdriver.F90:330]: Please add a comment describing what the created file will contain.

File: benchmarks/smallmatvec/manual_versions/openmp/kdriver.F90 [hunk 5]
```diff
@@ -0,0 +1,669 @@
...
+                call MVFUNCTION(VERSION)(cmap(colour,row) + i1, nlayers, &
+                    theta_adv_term_data, x_data, ncell_3d, ptheta_2_local_stencil, &
+                    ndf_any_space_1_theta_adv_term, undf_any_space_1_theta_adv_term, &
+                    map1, ndf_any_space_2_x, undf_any_space_2_x, map2 )
+              endif
+            enddo
+          end do
+          !$omp end parallel do
+        end do
+        !call exit(0)
```
Comment (arporter) [inline benchmarks/smallmatvec/manual_versions/openmp/kdriver.F90:582]: rm this line?

File: benchmarks/smallmatvec/manual_versions/openmp/matrix_vector_kernel_mod.F90 [hunk 6]
```diff
@@ -0,0 +1,668 @@
...
+    lhs_e = matmul(matrix(:,:,ik),x_e)
+    
+    do df = 1,ndf1
+       lhs(map1(df)+k) = lhs(map1(df)+k) + lhs_e(df) 
+    end do
+ end do
+ 
+end subroutine matrix_vector_code_original
+
+subroutine matrix_vector_code_fuseinterleaved(cell,        &
```
Comment (arporter) [inline benchmarks/smallmatvec/manual_versions/openmp/matrix_vector_kernel_mod.F90:112]: Please could you add a comment to each of the various implementations that briefly explains what it is that they test/do.

File: benchmarks/smallmatvec/manual_versions/openmp/problemsize.sh [hunk 7]
```diff
@@ -0,0 +1,33 @@
+#!/usr/bin/env bash
+
```
Comment (arporter) [inline benchmarks/smallmatvec/manual_versions/openmp/problemsize.sh:2]: Please add a comment that briefly describes what this script does.

File: benchmarks/smallmatvec/manual_versions/openmp/scalability.sh [hunk 8]
```diff
@@ -0,0 +1,29 @@
+#!/usr/bin/env bash
+
```
Comment (arporter) [inline benchmarks/smallmatvec/manual_versions/openmp/scalability.sh:2]: Please add a comment describing what this script does.

File: benchmarks/smallmatvec/manual_versions/openmp/utils.F90 [hunk 9]
```diff
@@ -0,0 +1,45 @@
+module utils
+    implicit none
+    contains
+subroutine system_mem_usage(valueRSS)
+    use ifport !if on intel compiler
```
Comment (arporter) [inline benchmarks/smallmatvec/manual_versions/openmp/utils.F90:5]: Please add a brief docstring, mentioning any limitations. (e.g. just ifort? and presumably only certain flavours of OS?_

File: benchmarks/smallmatvec/manual_versions/openmp/vectorization.sh [hunk 10]
```diff
@@ -0,0 +1,15 @@
+#!/usr/bin/env bash
+
```
Comment (arporter) [inline benchmarks/smallmatvec/manual_versions/openmp/vectorization.sh:2]: Please add a brief comment.

File: benchmarks/smallmatvec/README.md [hunk 11]
```diff
@@ -0,0 +1,13 @@
+# LFRic Small Matrix Vector Multiplications Benchmark
+
+This project contains multiple implementations of the matrix-vector
+multiplication operation done in the LFRic application. 
+
+## Manual Versions
+
+Manual implementations of the kernel are found in teh `manual_versions`
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply benchmarks/smallmatvec/README.md:8]: Fixed

File: benchmarks/smallmatvec/README.md [hunk 12]
```diff
@@ -0,0 +1,13 @@
...
+directory. At the moment the following versions are available:
+
+    - `openmp`: Configurable implementation using OpenMP that test
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply benchmarks/smallmatvec/README.md:11]: Fixed

File: benchmarks/smallmatvec/manual_versions/openmp/LICENNCE.MetOffice [hunk 13]
```diff
@@ -0,0 +1,30 @@
+Copyright (c) 2017,  Met Office, on behalf of HMSO and Queen's Printer
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply benchmarks/smallmatvec/manual_versions/openmp/LICENNCE.MetOffice:1]: Oh yes.

File: benchmarks/smallmatvec/manual_versions/openmp/Makefile [hunk 14]
```diff
@@ -0,0 +1,66 @@
...
+		$(FC) $(FFLAGS) -xAVX2 -DVERSION=$$v -o $@.$$v kdriver.F90 $^ $(LDFLAGS); \
+	done
+
+
+kdriver.novec: utils.o constants_mod.o argument_mod.o dino_mod.o kernel_mod.o matrix_vector_kernel_mod.novec.o 
+	for v in $(VERSIONS); do \
+		$(FC) $(FFLAGS) -no-vec -no-simd -DVERSION=$$v -o $@.$$v kdriver.F90 $^ $(LDFLAGS) ; \
+	done
+
+		#$(FC) $(FFLAGS) -no-vec -no-simd -I${HOME}/tools/libxsmm/include -DVERSION=$$v -DNEW_DL=1 -o $@.$$v kdriver.F90 $^ $(LDFLAGS) -mkl; \
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply benchmarks/smallmatvec/manual_versions/openmp/Makefile:51]: I removed this line here and gave some info at the top of the file about how to compile with MKL or libxsmm.

File: benchmarks/smallmatvec/manual_versions/openmp/README.md [hunk 15]
```diff
@@ -0,0 +1,32 @@
+# Configurable OpenMP manual implementation
+
+
+This implementation uses OpenMP and test the vectorization and threading
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply benchmarks/smallmatvec/manual_versions/openmp/README.md:4]: Fixed

File: benchmarks/smallmatvec/manual_versions/openmp/README.md [hunk 16]
```diff
@@ -0,0 +1,32 @@
...
+capabilities of multiple versions of the kernel and multiple parallel
+traversing strategies.
+
+Th most relevant files are:
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply benchmarks/smallmatvec/manual_versions/openmp/README.md:8]: Fixed

File: benchmarks/smallmatvec/manual_versions/openmp/README.md (hunk 1 above)
Comment (sergisiso) [reply benchmarks/smallmatvec/manual_versions/openmp/README.md:32]: Fixed

File: benchmarks/smallmatvec/manual_versions/openmp/README.md (hunk 2 above)
Comment (sergisiso) [reply benchmarks/smallmatvec/manual_versions/openmp/README.md:23]: Good idea

File: benchmarks/smallmatvec/manual_versions/openmp/innerloop.sh (hunk 3 above)
Comment (sergisiso) [reply benchmarks/smallmatvec/manual_versions/openmp/innerloop.sh:3]: Done

File: benchmarks/smallmatvec/manual_versions/openmp/kdriver.F90 (hunk 5 above)
Comment (sergisiso) [reply benchmarks/smallmatvec/manual_versions/openmp/kdriver.F90:582]: Done

File: benchmarks/smallmatvec/manual_versions/openmp/kdriver.F90 (hunk 4 above)
Comment (sergisiso) [reply benchmarks/smallmatvec/manual_versions/openmp/kdriver.F90:330]: I decided to delete this block, I used it to debug the generation of cell colours, but now this functionality already works and this code just works under certain conditions, so I don't think is needed anymore.

File: benchmarks/smallmatvec/manual_versions/openmp/matrix_vector_kernel_mod.F90 (hunk 6 above)
Comment (sergisiso) [reply benchmarks/smallmatvec/manual_versions/openmp/matrix_vector_kernel_mod.F90:112]: Done

File: benchmarks/smallmatvec/manual_versions/openmp/utils.F90 (hunk 9 above)
Comment (sergisiso) [reply benchmarks/smallmatvec/manual_versions/openmp/utils.F90:5]: Done

File: benchmarks/smallmatvec/manual_versions/openmp/matrix_vector_kernel_mod.F90 (hunk 6 above)
Comment (sergisiso) [reply benchmarks/smallmatvec/manual_versions/openmp/matrix_vector_kernel_mod.F90:112]: Done

File: benchmarks/smallmatvec/manual_versions/openmp/problemsize.sh (hunk 7 above)
Comment (sergisiso) [reply benchmarks/smallmatvec/manual_versions/openmp/problemsize.sh:2]: Done

File: benchmarks/smallmatvec/manual_versions/openmp/utils.F90 (hunk 9 above)
Comment (sergisiso) [reply benchmarks/smallmatvec/manual_versions/openmp/utils.F90:5]: Done

File: benchmarks/smallmatvec/manual_versions/openmp/scalability.sh (hunk 8 above)
Comment (sergisiso) [reply benchmarks/smallmatvec/manual_versions/openmp/scalability.sh:2]: Done

File: benchmarks/smallmatvec/manual_versions/openmp/vectorization.sh (hunk 10 above)
Comment (sergisiso) [reply benchmarks/smallmatvec/manual_versions/openmp/vectorization.sh:2]: Done

**Comment by sergisiso** [comment]
> Mostly it's just requests for comments so that we at least know what each file is for. 

Good idea, now that this is public it should be documented more thoroughly. I also created issue #27 for continue integrating this benchmark with the PSycloneBench infrastructure.

I did test that the Makefile works and basic executions of the binaries and scripts run as expected. (although it is just one set-up and I am sure there are corner cases that need to be polished)

Ready for next review.

File: benchmarks/smallmatvec/manual_versions/openmp/kdriver.F90 (hunk 4 above)
Comment (arporter) [reply benchmarks/smallmatvec/manual_versions/openmp/kdriver.F90]: Great.

File: benchmarks/smallmatvec/manual_versions/openmp/matrix_vector_kernel_mod.F90 [hunk 17]
```diff
@@ -211,6 +221,7 @@ subroutine matrix_vector_code_nlayersf(cell,        &
 
 end subroutine matrix_vector_code_nlayersf
 
+! Similar to nlayersf but
```
Comment (arporter) [inline benchmarks/smallmatvec/manual_versions/openmp/matrix_vector_kernel_mod.F90:224]: but...? :-)

File: benchmarks/smallmatvec/manual_versions/openmp/matrix_vector_kernel_mod.F90 [hunk 18]
```diff
@@ -368,7 +382,9 @@ subroutine matrix_vector_code_nlayersf_split(cell,        &
 end subroutine matrix_vector_code_nlayersf_split
 
 
-
+! Like the original but manually blocking the outer loop with blocks of size
+! 8 (vlen variable) to match the vector length. This this blocks are iterated
```
Comment (arporter) [inline benchmarks/smallmatvec/manual_versions/openmp/matrix_vector_kernel_mod.F90:386]: s/This this blocks are iterated/The members of each block are then iterated over in an inner loop/

File: benchmarks/smallmatvec/manual_versions/openmp/matrix_vector_kernel_mod.F90 (hunk 17 above)
Comment (sergisiso) [reply benchmarks/smallmatvec/manual_versions/openmp/matrix_vector_kernel_mod.F90]: oops, done now

File: benchmarks/smallmatvec/manual_versions/openmp/matrix_vector_kernel_mod.F90 (hunk 18 above)
Comment (sergisiso) [reply benchmarks/smallmatvec/manual_versions/openmp/matrix_vector_kernel_mod.F90]: Done

**Comment by sergisiso** [comment]
@arporter Thanks for catching this, in some of the implementations I didn't remember what they were doing so I skipped some and forgot to came back to it :)

**Comment by arporter** [comment]
Sorry Sergi but I've just realised that the top-level README still refers to the ocean directory. Please could you update and check the other READMEs.

**Comment by sergisiso** [comment]
@arporter Can you point me where are you referring?

**Comment by arporter** [comment]
> 
> 
> @arporter Can you point me where are you referring?

Actually, I've realised I was being stupid :-) It's all OK now.
